# eBay Auto Dataset Project

In this project I will work with 50,000 row sample from a dataset of German eBay car sales. The base dataset had been cleaned for analysis. For the purposes of this exercise, this sample has been "dirtied" again so that it resembles what one might find in a scraped dataset.


## Initial Data Dictionary

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import pprint as pp

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1" )

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Initial Observations

See above for details.

- Column names are in English.
- Data is in German
- Number formats are US-style '.' for decimal, ',' for thousands.
  - `price` format is $X,XXX (not Euros)
  - `odometer` format is XXX,XXXXkm
  - these were not recognized as numbers and columns have "object type.
- Timestamps in `dateCrawled` and `adCreated` and `lastSeen` weren't recognized as such.
- Some columns contain null values
  - `vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage`
- `notRepairedDamage` column has "NaN", "nein" & 'ja'

## Cleaning Column Names

In [5]:
autos.columns.tolist()

['dateCrawled',
 'name',
 'seller',
 'offerType',
 'price',
 'abtest',
 'vehicleType',
 'yearOfRegistration',
 'gearbox',
 'powerPS',
 'model',
 'odometer',
 'monthOfRegistration',
 'fuelType',
 'brand',
 'notRepairedDamage',
 'dateCreated',
 'nrOfPictures',
 'postalCode',
 'lastSeen']

Updating coulumn names from CamelCase to snakecase, and to be more descriptive

In [6]:
new_column_names = ['date_crawled',
                     'name',
                     'seller',
                     'offer_type',
                     'price',
                     'abtest',
                     'vehicle_type',
                     'registration_year',
                     'gearbox',
                     'power_ps',
                     'model',
                     'odometer',
                     'registration_month',
                     'fuel_type',
                     'brand',
                     'unrepaired_damage',
                     'ad_created',
                     'number_of_pictures',
                     'postal_code',
                     'last_seen']
autos.columns = new_column_names

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning.

In [8]:
autos.describe()

registration_year      power_ps  registration_month  \
count       50000.000000  50000.000000        50000.000000   
mean         2005.073280    116.355920            5.723360   
std           105.712813    209.216627            3.711984   
min          1000.000000      0.000000            0.000000   
25%          1999.000000     70.000000            3.000000   
50%          2003.000000    105.000000            6.000000   
75%          2008.000000    150.000000            9.000000   
max          9999.000000  17700.000000           12.000000   

       number_of_pictures   postal_code  
count             50000.0  50000.000000  
mean                  0.0  50813.627300  
std                   0.0  25779.747957  
min                   0.0   1067.000000  
25%                   0.0  30451.000000  
50%                   0.0  49577.000000  
75%                   0.0  71540.000000  
max                   0.0  99998.000000

In [9]:
autos.describe(include=np.object)

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   

       vehicle_type  gearbox  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       unrepaired_damage           ad_created            last_seen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

In [10]:
columns_of_interest = [
    "seller", 
    "offer_type", 
    "abtest", 
    "vehicle_type", 
    "gearbox", 
    "fuel_type", 
    "odometer", 
    "brand"
]

for c in columns_of_interest:
    print(c, ":\n{}\n".format(autos[c].value_counts()))

seller :
privat        49999
gewerblich        1
Name: seller, dtype: int64

offer_type :
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

abtest :
test       25756
control    24244
Name: abtest, dtype: int64

vehicle_type :
limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

gearbox :
manuell      36993
automatik    10327
Name: gearbox, dtype: int64

fuel_type :
benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

odometer :
150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

brand :
volkswagen        10687
opel 

### ToDo

Drop: 

* `seller` and `offer_type`, because all but one row have same value.  
* `number_of_photos`, because the value is 0 for all rows.
* `abtest` because we don't have enough information for it to be useful.

Fix: 

- Convert `price` and `odometer` to numeric values
- Rename `odometer` to `odometer_km`
- Maybe convert: `date_crawled` and `last_seen` to datetime

In [11]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [12]:
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)

In [13]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [14]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [15]:
autos.drop(columns=["seller", "offer_type", "number_of_pictures"], inplace=True)

In [16]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

## Exploring the Odometer and Price Columns

In [17]:
autos["odometer_km"].value_counts(dropna=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The above values seem reasonable. Nothing seems out of place

In [18]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [19]:
autos["price"].dtype
#Weird, considering what's reported by describe.

dtype('int64')

In [20]:
autos["price"].isnull().sum()

0

In [21]:
autos["price"].value_counts(dropna=False).sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

Some of these seem a little sketchy. 

How likely is it that a car sold for \\$99.9M? \\$27M seems like a stretch, but its possible.

Similarly, how likely is it that 1421 sold for \\$0, or 156 for \\$1? Does eBay even accept \\$0?

I am going to discard the 1421 rows with a price of $0.

In [22]:
autos = autos[autos["price"] > 0 ]
len(autos)

48579

Now lets look at the details of rows with a price >= \\$900K since there are only ~10.

In [23]:
autos[autos["price"] >= 900000]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

          price   abtest vehicle_type  registration_year    gearbox  power_ps  \
514      999999     test        kombi               2009    manuell       101   
2897   11111111     test    limousine               1973    manuell        48   
7814    1300000  control        coupe               1992        NaN         0   
11137  10000000  control        coupe               1960    manuell       368   
22947   1234566  control        kombi               1999  automatik       190   
24384  11111111     test          NaN               1995        NaN         0   
27371  12345678  control          NaN               2017        NaN        95   
37585    999990     test    limousine               1985    manuell       111   
39377  12345678  control          NaN               2018    manuell        95   
39705  99999999  control    limousine               1999  automatik       224   
42221  27322222  control    limousine               2014    manuell       163   
43049    999999     test          bus               1981    manuell        70   
47598  12345678  control    limousine               2001    manuell       101   
47634   3890000     test        coupe               2006        NaN       799   

             model  odometer_km  registration_month fuel_type           brand  \
514          focus       125000                   4    benzin            ford   
2897        escort        50000                   3    benzin            ford   
7814           NaN        50000                  12       NaN  sonstige_autos   
11137          NaN       100000                   1    benzin  sonstige_autos   
22947          NaN       150000                   2    diesel             bmw   
24384          NaN       150000                   0       NaN      volkswagen   
27371        punto       150000                   0       NaN            fiat   
37585        jetta       150000                  12    benzin      volkswagen   
39377          v40       150000                   6       NaN           volvo   
39705     s_klasse       150000                   9    benzin   mercedes_benz   
42221           c4        40000                   2    diesel         citroen   
43049  transporter       150000                   1    benzin      volkswagen   
47598       vectra       150000                   3    benzin            opel   
47634          NaN         5000                   7       NaN  sonstige_autos   

      unrepaired_damage           ad_created  postal_code            last_seen  
514                nein  2016-03-17 00:00:00        12205  2016-04-06 07:17:35  
2897               nein  2016-03-12 00:00:00        94469  2016-03-12 22:45:27  
7814               nein  2016-04-04 00:00:00        60598  2016-04-05 11:34:11  
11137              nein  2016-03-29 00:00:

The Ferrari prices are probably legitimate, as is the Maserati. The Fords, Fiat, VW Golf & Bus and BMW stationwagon probably aren't legitimate. The rest I don't know enough about.

Since there are only a few Ferraris in this price range, and since later analysis focuses on popular brands (which Ferarri isn't), I'm just going to drop everything above $900K.

In [24]:
autos = autos[autos["price"] < 900000]
len(autos)

48565

## Exploring the Date Columns

`date_crawled`, `ad_created` and `last_seen` all have timestamps, but are stored as string objects.

In [25]:
type(autos["date_crawled"].iloc[0])

str

In [26]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [27]:
for c in ["date_crawled", "ad_created", "last_seen"]:
    print(c,":")
    print(autos[c].str[:10].value_counts(normalize=True, dropna=False).sort_index().round(6)*100)
    print()

date_crawled :
2016-03-05    2.5327
2016-03-06    1.4043
2016-03-07    3.6014
2016-03-08    3.3296
2016-03-09    3.3090
2016-03-10    3.2184
2016-03-11    3.2575
2016-03-12    3.6920
2016-03-13    1.5670
2016-03-14    3.6549
2016-03-15    3.4284
2016-03-16    2.9610
2016-03-17    3.1628
2016-03-18    1.2911
2016-03-19    3.4778
2016-03-20    3.7887
2016-03-21    3.7373
2016-03-22    3.2987
2016-03-23    3.2225
2016-03-24    2.9342
2016-03-25    3.1607
2016-03-26    3.2204
2016-03-27    3.1092
2016-03-28    3.4860
2016-03-29    3.4099
2016-03-30    3.3687
2016-03-31    3.1834
2016-04-01    3.3687
2016-04-02    3.5478
2016-04-03    3.8608
2016-04-04    3.6487
2016-04-05    1.3096
2016-04-06    0.3171
2016-04-07    0.1400
Name: date_crawled, dtype: float64

ad_created :
2015-06-11    0.0021
2015-08-10    0.0021
2015-09-09    0.0021
2015-11-10    0.0021
2015-12-05    0.0021
               ...  
2016-04-03    3.8855
2016-04-04    3.6858
2016-04-05    1.1819
2016-04-06    0.3253
2016-04-07  

* `date_crawled` spans a ~one month range of dates from 2016/3/5 to 2016/4/7. From 0.3-3.8% 
* `ad_created` ranges from 2015/06/11 to 2016/04/07.
* `last_seen` ranges from 2016-03-05 to 2016-04-07

In [28]:
for c in ["date_crawled", "ad_created", "last_seen"]:
    print(c,":")
    print(pd.to_datetime(autos[c].str[:10]).describe(datetime_is_numeric=False))
    print()

date_crawled :
count                   48565
unique                     34
top       2016-04-03 00:00:00
freq                     1875
first     2016-03-05 00:00:00
last      2016-04-07 00:00:00
Name: date_crawled, dtype: object

ad_created :
count                   48565
unique                     76
top       2016-04-03 00:00:00
freq                     1887
first     2015-06-11 00:00:00
last      2016-04-07 00:00:00
Name: ad_created, dtype: object

last_seen :
count                   48565
unique                     34
top       2016-04-06 00:00:00
freq                    10772
first     2016-03-05 00:00:00
last      2016-04-07 00:00:00
Name: last_seen, dtype: object



<ipython-input-28-91074581b22a>:3: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  print(pd.to_datetime(autos[c].str[:10]).describe(datetime_is_numeric=False))
<ipython-input-28-91074581b22a>:3: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  print(pd.to_datetime(autos[c].str[:10]).describe(datetime_is_numeric=False))
<ipython-input-28-91074581b22a>:3: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior n

It seems strange that out of 50,000 listings, created over a span of almost a year, there are only 76 unique creation dates. On further consideration, it makes sense. The ads themsevles were  gathered over a period of one month and most of the ads were created within or shortly before that period. The older dates are increacingly sparse and represent only cars that have taken a long time to sell.

The range in the number of pages crawled on a given day is also interesting. It varies over a pretty wide range. Thinking it through, it's may be a product of both the rate of the spider and the rate of ad creation over time. If the spider can't crawl all the ads in a given run, it may not detect a new ad for some period of time. Ads may also be missed if they end before detection.

In [29]:
autos[["date_crawled", "ad_created", "last_seen"]] = autos[["date_crawled", "ad_created", "last_seen"]].astype('datetime64')

## Analyzing and Cleaning Registration Year Data

In [30]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [31]:
autos["registration_year"].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

In [32]:
autos.loc[autos["registration_year"].between(1900,2016),"registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

There are a number of entries with bogus registration years. I've eliminated anything more recent than 2016, when the crawls were performed. It would be more precise to compare the full ad creation date to year and month of registration, but this is good enough to get a sense of the data.

I've also elimated anything with a registration date earlier than 1900. This seems on the one hand, a bit unnecessary, since it eliminates less than a dozen listings, while also possibly a bit lax, since I don't know when Germany first started registering cars.

The average car sold was ~15y old, and 75% of cars were at least ~8y old.

In [33]:
autos.loc[autos["registration_year"] < 1935,
          ["name", "model", "brand", "registration_year", "registration_month"]
         ]

name    model  \
2221   Sehr_seltener_Oldtimer_Opel_1210_zum_Restaurieren   andere   
2573               Hanomag_rekord_15k_Suche_ersatz_teile      NaN   
3679                                          Suche_Auto      NaN   
10556                                         UNFAL_Auto      NaN   
11246                  Ford_Model_A_Roadster_Deluxe_1931   andere   
21416                            Essex_super_six__Ford_A   andere   
22101                                         BMW_Andere   andere   
22316              VW_Kaefer.__Zwei_zum_Preis_von_einem.   kaefer   
22659                                       Opel_Corsa_B    corsa   
24511                        Trabant__wartburg__Ostalgie      NaN   
28693                                     Renault_Twingo      NaN   
30781                   Opel_Calibra_V6_DTM_Bausatz_1:24  calibra   
32585                                         UNFAL_Auto      NaN   
45157                                         Motorhaube      NaN   
49283                                         Citroen_HY   andere   

                brand  registration_year  registration_month  
2221             opel               1934                   0  
2573   sonstige_autos               1934                   1  
3679   sonstige_autos               1910                   0  
10556      mitsubishi               1800                   2  
11246            ford               1931                   7  
21416            ford               1927                   5  
22101             bmw               1929                   1  
22316      volkswagen               1000                   0  
22659            opel               1910                   0  
24511         trabant               1111                   0  
28693         renault               1910                   0  
30781            opel               1910                   0  
32585      mitsubishi               1800                   2  
45157         trabant               1910                   0  
49283         citroen               1001                   0

To my eye, the earliest genuine registration is a 1927 Ford Model A. The next is a 1931 Ford Model A. In between the two there is a BMW Andere (Other) registered 1929/1. There are some other early 30s listings that appear to be old cars. I'm going to pick 1927 as the earliest legitimate registration year.

In [34]:
autos.loc[autos["registration_year"] > 2016,
          ["name", "model", "brand", "registration_year", "registration_month"]
         ]

name     model  \
10                          VW_Golf_Tuning_in_siber/grau       NaN   
55                        Mercedes_E320_AMG_zu_Tauschen!  e_klasse   
65                         Ford_Fiesta_zum_ausschlachten    fiesta   
68     Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   clubman   
84                                        Renault_twingo    twingo   
...                                                  ...       ...   
49770                     VW_Polo_6n_Tuev_Neu!__1.6_75PS      polo   
49796                        Opel_corsa_1.4_zu_verkaufen     corsa   
49841                                  Passat_abzugeben.    passat   
49910  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...    andere   
49935  Mercedes_A_klasse_angemeldet_mit_Tuef_und_Auto...  a_klasse   

               brand  registration_year  registration_month  
10        volkswagen               2017                   4  
55     mercedes_benz               2017                   7  
65              ford               2017                   9  
68              mini               2017                   0  
84           renault               2018                   0  
...              ...                ...                 ...  
49770     volkswagen               2018                  12  
49796           opel               2017                   7  
49841     volkswagen               2017                   7  
49910           opel               9000                   3  
49935  mercedes_benz               2017                   9  

[1879 rows x 5 columns]

In [35]:
autos.loc[autos["registration_year"] > 2016, "registration_year"].value_counts()

2017    1392
2018     470
5000       4
9999       3
2019       2
5911       1
9000       1
2800       1
4800       1
8888       1
4500       1
6200       1
4100       1
Name: registration_year, dtype: int64

There are 1453 entries with a registration date of 2017, and 492 with a registration date of 2018. This makes me think it may be worthwhile to look within 2016 for bogus registration dates. The easiest is to look for anything with a registration month more recent than 4/2016, since that's when the last crawl was done.

In [36]:
autos.loc[(autos["registration_year"] == 2016) & (autos["registration_month"] > 4),
         ["name", "model", "brand", "registration_year", "registration_month"]]

name     model  \
135    Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...    meriva   
256                       Passat_1.9TDI_4Motion_Highline    passat   
295                                       Privat_anbiter     astra   
307                                         Giessen_ford     focus   
437                        Mazda__klima_leder__Alufelgen       NaN   
...                                                  ...       ...   
49547                    Smart_Passion_mit_Panorama_Dach    fortwo   
49852                                   TOP__Golf_3_1.8l      golf   
49876                             Audi_a5_3.0_tdi_s_line       NaN   
49919                                         Fiat_Punto     punto   
49938                     Mercedes_Benz_A_160_Avantgarde  a_klasse   

               brand  registration_year  registration_month  
135             opel               2016                   8  
256       volkswagen               2016                  11  
295             opel               2016                   8  
307             ford               2016                   8  
437            mazda               2016                   7  
...              ...                ...                 ...  
49547          smart               2016                  10  
49852     volkswagen               2016                   7  
49876           audi               2016                  10  
49919           fiat               2016                   8  
49938  mercedes_benz               2016                   6  

[618 rows x 5 columns]

In [37]:
autos = autos[autos["registration_year"].between(1917,2016)]
autos = autos[~((autos["registration_year"] == 2016) & (autos["registration_month"] > 4))]

In [38]:
autos["registration_year"].value_counts(normalize=True).round(1)*100

2000    10.0
2005    10.0
1999    10.0
2004    10.0
2003    10.0
        ... 
1939     0.0
1927     0.0
1929     0.0
1948     0.0
1952     0.0
Name: registration_year, Length: 77, dtype: float64

In [39]:
autos["registration_year"].describe()

count    46058.000000
mean      2002.745213
std          7.004326
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2016.000000
Name: registration_year, dtype: float64

I've eliminated 2638 entries with problematic registration years. My observations after my first refinement of the dataset still holds:

The average car sold was ~15y old, and 75% of cars were at least ~8y old.

In [40]:
len(autos)

46058

## Top Brands

In [41]:
autos["brand"].describe()

count          46058
unique            40
top       volkswagen
freq            9708
Name: brand, dtype: object

In [42]:
autos["brand"].value_counts(normalize=True).round(2)*100

volkswagen        21.0
bmw               11.0
opel              11.0
mercedes_benz     10.0
audi               9.0
ford               7.0
renault            5.0
peugeot            3.0
fiat               3.0
seat               2.0
skoda              2.0
nissan             2.0
mazda              2.0
smart              1.0
citroen            1.0
toyota             1.0
hyundai            1.0
sonstige_autos     1.0
volvo              1.0
mini               1.0
mitsubishi         1.0
honda              1.0
kia                1.0
alfa_romeo         1.0
porsche            1.0
suzuki             1.0
chevrolet          1.0
chrysler           0.0
dacia              0.0
daihatsu           0.0
jeep               0.0
subaru             0.0
land_rover         0.0
saab               0.0
jaguar             0.0
daewoo             0.0
trabant            0.0
rover              0.0
lancia             0.0
lada               0.0
Name: brand, dtype: float64

In [43]:
top10_counts = autos["brand"].value_counts(normalize=True)[:10]
top10_counts.sum()

0.8024664553389205

## Exploring Price By Brand

There are forty unique auto brands represented in the dataset. The top four brands account for ~50% of the entries. Three are German domestic manufacturers, the fourth is Opel, which is GM's German subsidiary. The fifth most common brand is Audi, also German. The top five account for ~60% of the cars in the dataset.

The top ten brands account for 80% of the dataset.

I am going to look at some key aggregate details for the top ten brands.

### Brands and Average Prices

In [44]:
top_brands = [brand for brand in top10_counts.index]
top_brands

['volkswagen',
 'bmw',
 'opel',
 'mercedes_benz',
 'audi',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'seat']

In [45]:
brand_mean_prices = {}
for brand in top_brands:
    selection = autos.loc[autos["brand"] == brand]
    brand_mean_prices[brand] = selection["price"].mean()

brand_mean_prices = pd.Series(brand_mean_prices)
brand_mean_prices

volkswagen       5444.281314
bmw              8367.863788
opel             2996.046644
mercedes_benz    8681.313299
audi             9387.175119
ford             3790.819059
renault          2493.956181
peugeot          3108.575492
fiat             2830.815946
seat             4443.226347
dtype: float64

In [46]:
sum(brand_mean_prices)/len(brand_mean_prices)

5154.407318935544

In [47]:
autos["price"].describe()

count     46058.000000
mean       6022.932954
std        9219.593098
min           1.000000
25%        1250.000000
50%        3150.000000
75%        7600.000000
max      350000.000000
Name: price, dtype: float64

* The mean price for all cars in the dataset is \\$6,022. (It was \\$9,685 before removing (~10) cars above \\$900,000 from the dataset)

* The mean price for cars from the top ten brands is \\$5,154

* The mean price of cars from the top-ten brands is below the dataset mean.

* Among USA-based manufactuers, the average Ford sells for 26% more than the average Opel (GM).

* Among the top tend brands, Audi, Mercedes and BMW cars command a premium. Volkswagon is right in the middle and Fiat, Ford, Opel, Peugot and Renault are budget brands.

## Relationship between Price and Odometer Reading

In [48]:
brand_mean_odometer = {}
for brand in top_brands:
    selection = autos.loc[autos["brand"] == brand]
    brand_mean_odometer[brand] = selection["odometer_km"].mean()

brand_mean_odometer = pd.Series(brand_mean_odometer)
brand_mean_odometer

volkswagen       128550.164813
bmw              132490.677134
opel             129196.917461
mercedes_benz    130638.035434
audi             129056.957282
ford             124051.697291
renault          127984.317343
peugeot          126867.250182
fiat             116849.024597
seat             120838.323353
dtype: float64

In [49]:
brand_data = pd.DataFrame(brand_mean_prices, columns=['mean_price'])

In [50]:
brand_data["mean_odometer"] = brand_mean_odometer

Calculate $ per 1,000km to allow easier comparison among brands.

In [51]:
brand_data["odometer_normalized_price"] = 1000*brand_data['mean_price']/brand_data['mean_odometer']

In [52]:
brand_data.round(0)

mean_price  mean_odometer  odometer_normalized_price
volkswagen         5444.0       128550.0                       42.0
bmw                8368.0       132491.0                       63.0
opel               2996.0       129197.0                       23.0
mercedes_benz      8681.0       130638.0                       66.0
audi               9387.0       129057.0                       73.0
ford               3791.0       124052.0                       31.0
renault            2494.0       127984.0                       19.0
peugeot            3109.0       126867.0                       25.0
fiat               2831.0       116849.0                       24.0
seat               4443.0       120838.0                       37.0

In [53]:
brand_data.describe()

mean_price  mean_odometer  odometer_normalized_price
count    10.000000      10.000000                  10.000000
mean   5154.407319  126652.336489                  40.343285
std    2676.064597    4764.710803                  20.011273
min    2493.956181  116849.024597                  19.486420
25%    3024.178856  124755.585514                  24.295348
50%    4117.022703  128267.241078                  33.664196
75%    7636.968170  129161.927416                  57.956453
max    9387.175119  132490.677134                  72.736684

* Differences in average odometer reading among top brands does not explain differences in average selling prices.
* Standard deviation for mean_price among brands is ~50% the mean, while the standard deviation for odometer readings among brands is <4% the mean.

It would be interesting to know the original selling price of the vehicles, so one could look at % depreciation vs odometer reading.

## Translation of German Terms

Most of the categorical columns contain German words. In order to prepare for further analysis, I will identify those values and replace them with english equivalents.

In [54]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage ad_created  postal_code           last_seen  
0              nein 2016-03-26        79588 2016-04-06 06:45:54  
1              nein 2016-04-04        71034 2016-04-06 14:45:08  
2              nein 2016-03-26        35394 2016-04-06 20:15:37  
3              nein 2016-03-12        33729 2016-03-15 03:16:28  
4              nein 2016-04-01        39218 2016-04-01 14:38:50

In [55]:
# Generate some dictionary definitions that I can then cut, paste, edit and execute.
to_translate =["vehicle_type", "gearbox", "fuel_type", "unrepaired_damage"]
for c in to_translate:
    values = autos[c].unique()
    temp_dict = {v:v for v in values}
    print(c," = ", end="")
    pp.pprint(temp_dict)

vehicle_type  = {nan: nan,
 'andere': 'andere',
 'bus': 'bus',
 'cabrio': 'cabrio',
 'coupe': 'coupe',
 'kleinwagen': 'kleinwagen',
 'kombi': 'kombi',
 'limousine': 'limousine',
 'suv': 'suv'}
gearbox  = {nan: nan, 'automatik': 'automatik', 'manuell': 'manuell'}
fuel_type  = {nan: nan,
 'andere': 'andere',
 'benzin': 'benzin',
 'cng': 'cng',
 'diesel': 'diesel',
 'elektro': 'elektro',
 'hybrid': 'hybrid',
 'lpg': 'lpg'}
unrepaired_damage  = {nan: nan, 'ja': 'ja', 'nein': 'nein'}


In [56]:
vehicle_type_translation = {'andere': 'other',
                 'bus': 'bus',
                 'cabrio': 'convertable',
                 'coupe': 'coupe',
                 'kleinwagen': 'subcompact',
                 'kombi': 'stationwagon',
                 'limousine': 'limousine',
                 'suv': 'suv'}

autos["vehicle_type"] = autos["vehicle_type"].map(vehicle_type_translation)

In [57]:
gearbox_translation = {'automatik': 'automatic', 'manuell': 'manual'}
autos["gearbox"] = autos["gearbox"].map(gearbox_translation)

In [58]:
fuel_type_translation = {'andere': 'other',
              'benzin': 'gasoline',
              'cng': 'cng',
              'diesel': 'diesel',
              'elektro': 'electric',
              'hybrid': 'hybrid',
              'lpg': 'lpg'}

autos["fuel_type"] = autos["fuel_type"].map(fuel_type_translation)

In [59]:
unrepaired_damage_translation = {'ja': 'yes', 'nein': 'no'}
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired_damage_translation)

In [60]:
autos["model"] = autos["model"].str.replace("andere","other")

In [61]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest  vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control           bus               2004     manual       158   
1   8500  control     limousine               1997  automatic       286   
2   8990     test     limousine               2009     manual       102   
3   4350  control    subcompact               2007  automatic        71   
4   1350     test  stationwagon               2003     manual         0   

    model  odometer_km  registration_month fuel_type       brand  \
0   other       150000                   3       lpg     peugeot   
1     7er       150000                   6  gasoline         bmw   
2    golf        70000                   7  gasoline  volkswagen   
3  fortwo        70000                   6  gasoline       smart   
4   focus       150000                   7  gasoline        ford   

  unrepaired_damage ad_created  postal_code           last_seen  
0                no 2016-03-26        79588 2016-04-06 06:45:54  
1                no 2016-04-04        71034 2016-04-06 14:45:08  
2                no 2016-03-26        35394 2016-04-06 20:15:37  
3                no 2016-03-12        33729 2016-03-15 03:16:28  
4                no 2016-04-01        39218 2016-04-01 14:38:50

## Extracting Structured Data from Name Column

In [62]:
autos["name"]

0                         Peugeot_807_160_NAVTECH_ON_BOARD
1               BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik
2                               Volkswagen_Golf_1.6_United
3        Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...
4        Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...
                               ...                        
49995     Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon
49996    Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...
49997                       Fiat_500_C_1.2_Dualogic_Lounge
49998                   Audi_A3_2.0_TDI_Sportback_Ambition
49999                                  Opel_Vectra_1.6_16V
Name: name, Length: 46058, dtype: object

It might be practical to extract engine engine details, like displacement and presence of a high-performance valvetrain (ie "16V") from the name column in some of the rows. Let's see how easy it is to extract displacement.

In [63]:
displacements = autos["name"].str.extract(r'_(\d[_\.]\d+)_')
displacements = displacements.iloc[:,0].str.replace('_', '.')
displacements.value_counts().head(20)

2.0    3398
1.6    2364
1.4    2011
1.9    1680
1.2    1555
1.8    1258
3.0     583
1.0     546
2.5     544
2.2     466
1.3     332
1.7     187
1.5     185
2.4     159
2.7     157
1.1     148
2.8     137
2.3      95
3.2      86
4.2      77
Name: 0, dtype: int64

In [64]:
# Count non-null values
displacements.count()

16525

A simple regular expression extracts displacement values for slightly over 1/3rd of the dataset. A cursory examination shows that most of these values are plausible.

This could be used to enhance the dataset. How to use the enhancement is left as an exercise for the reader.

## Most Common Models by Brand

In [65]:
brands = np.sort(autos["brand"].unique())
for b in brands:
    selection = (autos["brand"] == b)
    print(b,":")
    print(autos.loc[selection, "model"].value_counts(normalize=True)[:3])
    print()

alfa_romeo :
156      0.297945
147      0.267123
other    0.202055
Name: model, dtype: float64

audi :
a4    0.314763
a3    0.207445
a6    0.203594
Name: model, dtype: float64

bmw :
3er    0.530892
5er    0.230974
1er    0.105565
Name: model, dtype: float64

chevrolet :
other    0.588710
matiz    0.133065
spark    0.096774
Name: model, dtype: float64

chrysler :
other        0.339869
voyager      0.274510
ptcruiser    0.196078
Name: model, dtype: float64

citroen :
other       0.350081
berlingo    0.155592
c5          0.116694
Name: model, dtype: float64

dacia :
sandero    0.361345
logan      0.352941
duster     0.226891
Name: model, dtype: float64

daewoo :
matiz    0.333333
kalos    0.257576
lanos    0.151515
Name: model, dtype: float64

daihatsu :
cuore     0.548673
other     0.123894
sirion    0.097345
Name: model, dtype: float64

fiat :
punto    0.360889
other    0.202667
500      0.105778
Name: model, dtype: float64

ford :
focus     0.243673
fiesta    0.229721
mondeo    0.1537

Above we can see the three most common models for each brand. In many cases "other" is one of the most common values, which suggests deficiencies in the tools for creating listings that eBay gives sellers.

I'm going to add a new column to the dataset that combines brand and model for use in later analysis.

In [66]:
autos["brand_model"] = autos["brand"] + "/" + autos["model"]
autos["brand_model"].value_counts().head(20)

volkswagen/golf           3640
bmw/3er                   2595
volkswagen/polo           1572
opel/corsa                1562
volkswagen/passat         1338
opel/astra                1331
audi/a4                   1226
bmw/5er                   1129
mercedes_benz/c_klasse    1129
mercedes_benz/e_klasse     950
audi/a3                    808
audi/a6                    793
ford/focus                 751
ford/fiesta                708
volkswagen/transporter     670
renault/twingo             602
peugeot/2_reihe            592
smart/fortwo               538
opel/vectra                534
mercedes_benz/a_klasse     524
Name: brand_model, dtype: int64

## Relationship between Price and Odometer Readings

An investigation of the relationship between selling price and odometer readings will have to take into account the wide range in average selling prices between the top ten brands. So, I've aggregated them by brand.

NOTE: Odometer readings in the dataset have already been binned, so no further binning has been performed.

In [67]:
for b in sorted(top_brands):
    brand_subset = autos[autos["brand"] == b]
    odometer_readings = np.sort(brand_subset["odometer_km"].unique())
    print(b,":")
    print(brand_subset[["price", "odometer_km"]].mean())
    for r in odometer_readings:
        odometer_subset = brand_subset[brand_subset["odometer_km"] == r]
        print(r,":", round(odometer_subset["price"].mean(),-3))
    price_40kkm = brand_subset.loc[brand_subset["odometer_km"] == 40000, "price"].mean()
    price_90kkm = brand_subset.loc[brand_subset["odometer_km"] == 90000, "price"].mean()
    print("90K/40K price ratio:",
         round(price_90kkm/price_40kkm,1))
    print()


audi :
price            9387.175119
odometer_km    129056.957282
dtype: float64
5000 : 16000.0
10000 : 28000.0
20000 : 27000.0
30000 : 27000.0
40000 : 26000.0
50000 : 23000.0
60000 : 20000.0
70000 : 22000.0
80000 : 16000.0
90000 : 16000.0
100000 : 15000.0
125000 : 12000.0
150000 : 6000.0
90K/40K price ratio: 0.6

bmw :
price            8367.863788
odometer_km    132490.677134
dtype: float64
5000 : 10000.0
10000 : 31000.0
20000 : 27000.0
30000 : 24000.0
40000 : 25000.0
50000 : 24000.0
60000 : 21000.0
70000 : 17000.0
80000 : 16000.0
90000 : 15000.0
100000 : 13000.0
125000 : 11000.0
150000 : 6000.0
90K/40K price ratio: 0.6

fiat :
price            2830.815946
odometer_km    116849.024597
dtype: float64
5000 : 5000.0
10000 : 10000.0
20000 : 7000.0
30000 : 8000.0
40000 : 7000.0
50000 : 5000.0
60000 : 5000.0
70000 : 5000.0
80000 : 4000.0
90000 : 4000.0
100000 : 3000.0
125000 : 2000.0
150000 : 2000.0
90K/40K price ratio: 0.5

ford :
price            3790.819059
odometer_km    124051.697291
dt

As an alternative, lets display the aggreagated data as a data frame.

In [68]:
brand_data['90_40_price_ratio'] = np.NAN  #done to place early in column order
odometer_readings = np.sort(autos["odometer_km"].unique())    
for r in odometer_readings:
    odometer_subset = autos[autos["odometer_km"] == r]
    for b in brand_data.index:
        brand_subset = odometer_subset[odometer_subset["brand"] == b]
        brand_data.loc[b,r] = round(brand_subset["price"].mean(),-3)

price_ratios = {}
for b in brand_data.index:
    price_ratios[b] = round(brand_data.loc[b,90000]/brand_data.loc[b,40000],2)
    
price_ratios = pd.Series(price_ratios)
brand_data["90_40_price_ratio"] = price_ratios

In [69]:
brand_data

mean_price  mean_odometer  odometer_normalized_price  \
volkswagen     5444.281314  128550.164813                  42.351415   
bmw            8367.863788  132490.677134                  63.158133   
opel           2996.046644  129196.917461                  23.189769   
mercedes_benz  8681.313299  130638.035434                  66.453183   
audi           9387.175119  129056.957282                  72.736684   
ford           3790.819059  124051.697291                  30.558381   
renault        2493.956181  127984.317343                  19.486420   
peugeot        3108.575492  126867.250182                  24.502584   
fiat           2830.815946  116849.024597                  24.226269   
seat           4443.226347  120838.323353                  36.770010   

               90_40_price_ratio     5000    10000    20000    30000    40000  \
volkswagen                  0.56   4000.0  19000.0  17000.0  15000.0  16000.0   
bmw                         0.60  10000.0  31000.0  27000.0  24000.0  25000.0   
opel                        0.44   4000.0  12000.0   9000.0  11000.0   9000.0   
mercedes_benz               0.57  11000.0  29000.0  28000.0  19000.0  28000.0   
audi                        0.62  16000.0  28000.0  27000.0  27000.0  26000.0   
ford                        0.40   8000.0  16000.0  12000.0  13000.0  10000.0   
renault                     0.57   5000.0   8000.0   9000.0  12000.0   7000.0   
peugeot                     0.40   3000.0  13000.0   5000.0  12000.0  10000.0   
fiat                        0.57   5000.0  10000.0   7000.0   8000.0   7000.0   
seat                        0.50   6000.0  20000.0  11000.0  15000.0  12000.0   

                 50000    60000    70000    80000    90000   100000   125000  \
volkswagen     13000.0  13000.0  12000.0  11000.0   9000.0   8000.0   6000.0   
bmw            24000.0  21000.0  17000.0  16000.0  15000.0  13000.0  11000.0   
opel            9000.0   7000.0   8000.0   6000.0   4000.0   5000.0   3000.0   
mercedes_benz  24000.0  22000.0  17000.0  16000.0  16000.0  14000.0  10000.0   
audi           23000.0  20000.0  22000.0  16000.0  16000.0  15000.0  12000.0   
ford           10000.0   7000.0   8000.0   6000.0   4000.0   5000.0   4000.0   
renault         7000.0   7000.0   6000.0   4000.0   4000.0   3000.0   2000.0   
peugeot         8000.0   6000.0   7000.0   5000.0   4000.0   5000.0   4000.0   
fiat            5000.0   5000.0   5000.0   4000.0   4000.0   3000.0   2000.0   
seat           10000.0  10000.0   8000.0   7000.0   6000.0   5000.0   4000.0   

               150000  
volkswagen     4000.0  
bmw            6000.0  
opel           2000.0  
mercedes_benz  6000.0  
audi           6000.0  
ford           2000.0  
renault        2000.0  
peugeot        2000.0  
fiat           2000.0  
seat           2000.0

Notice that the prices for the lowest mileage cars are unusually low, lower than for higher mileage cars. One hypothesis is that these tend to be wrecked cars with unrepaired damage. This could be investigated with the available data; I will instead undertake a different analysis of the prices of damaged vs undamaged cars later in this notebook.

I've experimented with finding a quick metric for comparing depreciation among brands. For each brand, I've computed the ratio of mean price for 40,000km vehicles with those for 90,000km.

Between 40,000km and 90,000km:
* Audi and BMW cars loose ~40% of their value 
* Mercedes, Volkswagon and Renault loose ~43%
* Fiat, Ford, Open, Peugeot loose >55% of their value

## Damaged Cars

In [70]:
damaged = autos[autos["unrepaired_damage"] == "yes"]
undamaged = autos[autos["unrepaired_damage"] == "no"]

In [71]:
print("Mean Damaged Price:", round(damaged["price"].mean(),-1))
print("Mean Undamaged Price:", round(undamaged["price"].mean(),-1))

Mean Damaged Price: 2260.0
Mean Undamaged Price: 7200.0


Cars with unrepaired damage sell, on average, for less than 1/3 the price of cars without unrepaired damage.

We can also look at price differences for damaged cars on a model by model basis.

In [72]:
autos.groupby(["brand_model","unrepaired_damage"])["price"].mean()

brand_model     unrepaired_damage
alfa_romeo/145  no                    2750.000000
alfa_romeo/147  no                    2722.854545
                yes                   1286.846154
alfa_romeo/156  no                    1689.516667
                yes                   1190.388889
                                         ...     
volvo/v60       no                   21000.000000
volvo/v70       no                    5009.258065
                yes                   1570.615385
volvo/xc_reihe  no                   13378.673913
                yes                  11450.000000
Name: price, Length: 524, dtype: float64

Another way of doing it...

In [73]:
damaged_comparison = pd.DataFrame(undamaged.groupby("brand_model")["price"].mean())
damaged_comparison.rename({"price": "mean_undamaged_price"}, axis=1, inplace=True)
damaged_comparison["mean_damaged_price"] = damaged.groupby("brand_model")["price"].mean()
damaged_comparison

mean_undamaged_price  mean_damaged_price
brand_model                                               
alfa_romeo/145             2750.000000                 NaN
alfa_romeo/147             2722.854545         1286.846154
alfa_romeo/156             1689.516667         1190.388889
alfa_romeo/159             6659.653846         7800.000000
alfa_romeo/other           8204.046512         3548.625000
...                                ...                 ...
volvo/v40                  2227.868852          756.461538
volvo/v50                  5545.375000         3495.000000
volvo/v60                 21000.000000                 NaN
volvo/v70                  5009.258065         1570.615385
volvo/xc_reihe            13378.673913        11450.000000

[287 rows x 2 columns]

In [74]:
damaged_comparison.sample(20)

mean_undamaged_price  mean_damaged_price
brand_model                                                     
saab/9000                        2912.250000          200.000000
seat/leon                        8999.803922         3861.461538
land_rover/discovery            15154.363636         1725.000000
kia/sportage                    12591.702128         1200.000000
alfa_romeo/147                   2722.854545         1286.846154
honda/civic                      4067.247706          846.812500
rover/other                      1304.212121          820.555556
land_rover/range_rover          25411.428571         1800.000000
seat/other                       1612.727273          399.666667
jaguar/x_type                   10132.000000         1424.750000
renault/espace                   3495.120000          648.333333
mercedes_benz/v_klasse          15362.250000         3512.250000
volkswagen/sharan                6684.231707         5235.214286
volkswagen/transporter          11924.129225         4932.614035
trabant/other                    3009.800000                 NaN
kia/sorento                      9139.153846         3549.800000
mercedes_benz/other             11804.480874         4206.064516
audi/q7                         22040.184211        11950.000000
mercedes_benz/clk                6881.747368         2981.538462
ford/ka                          1806.710000          472.270833

Not surprisingly damaged cars tend to sell for significantly less than undamaged cars. However, this is not universally true. Deeper analysis could be undertaken to try to explain the unexpected results. One hypothesis is that there are a relatively small number of cars in many of the make/model categories and that the comparison may be skewed by differences in odometer reading and other factors, including the extent of the damage.

## In Summary

* The dataset included a number of unusual values in a variety of columns. These values were investigated, in context, and some rows containing implausible values were dropped. In the end we retained 46,058 records out of the initial 50,000.
  * Dropped rows with prices higher than \\$900K
  * Dropped rows with registration years more recent than 2016, which is when the crawl was performed, and earlier than 1927 which I judged, based on the content of listings, to be the most recent legitimate registration year.
  *
* The average car sold was ~15 years old.
* 75\% of cars sold were at least 8 years old.
* The mean price for cars in the dataset (after clearning) is \\$6,002.
* The five most common brands are, in order Volkswagon, BMW, Opel, Mercedes Benz, and Audi.
* - Audi, BMW and Mercedes command a premium.

Reccomendations:

This dataset has some holes and inconsistences, some of which point to deficiencies in eBay's listing-creation tools.

One prominent example is that "Other" is one of the top-3 models for many car brands. Accurately capturing the model of the car is important for sellers, buyers and anyone else drawing on the data. Correctly capturing the model helps sellers establish a good estimate of the market price for their vehicle and that it reaches the largest number of potential buyers. Recording the correct model helps buyers find all the cars that meet their requirements and interests.